### DISCUSSION TITLE
Use a string literal value in a model association

### DISCUSSION THREAD

**Discussion by TSMMark** [jeremyevans/sequel#2234] [General]
Setting aside db schema design concerns... is it possible to define a model association where the FK value is a string literal defined in Ruby at definition time? 

I tried this but i think it is trying to call a method by name of the value provided of slug:

```base.many_to_one :network, key: Sequel.lit(base.slug), primary_key: :slug```

e.g. `slug = "youtube"`

```
undefined method `youtube' 
```

**Comment by jeremyevans** [comment]
By default, for normal association loading, Sequel calls methods on the receiver to get the values used in the association query.  If you want a static string literal, you could try defining a method that returns it.  I'll admit I'm not sure what value there would be in having an association defined that way, unless you were using some other approach to make the association receiver-specific.  Maybe it would help to provide a more complete example showing what you want?

**Comment by TSMMark** [comment]
Thanks for your quick and valuable response as always @jeremyevans 

> try defining a method that returns it

That sounds reasonable! But would that work for dataset things, like `association_join`?

I'll post a more complete example in a moment

**Comment by TSMMark** [comment]
Simplified the schemas and naming for the example, but it's something like this:

| networks | |
| --- | --- |
| id | PK |
| slug | unique slug consistent across environments |
| ... | |

| youtube_accounts | |
| --- | --- |
| id | PK |
| ... | |

| facebook_accounts | |
| --- | --- |
| id | PK |
| ... | |

We want YoutubeAccount many_to_one Network on a constant value "youtube", which is the value of `slug` for the network record corresponding to youtube.
And FacebookAccount many_to_one Network on constant "facebook".
Etc, for several more `*_account` tables

**Reply by jeremyevans** [reply]
Does the pk in `youtube_accounts` match `networks`? Otherwise, I don't think there is enough information, since the foreign key information is not provided.

You could also try posting the SQL you want the association query to generate.

**Reply by TSMMark** [reply]
The FK is essentially the slug. There is 1 record in `networks` for each slug; each `*accounts` model should be associated with 1 static slug. Will post example SQL generated next

**Reply by TSMMark** [reply]
I think the SQL would be pretty simple, just `WHERE` on the slug value for that model:

```
YoutubeAccount.sample.network # => #<Network @values={:id=>1, :slug=>"youtube"}>
# => SELECT * FROM "networks" WHERE ("slug" = 'youtube') LIMIT 1
```

```
FacebookAccount.sample.network # => #<Network @values={:id=>2, :slug=>"facebook"}>
# => SELECT * FROM "networks" WHERE ("slug" = 'facebook') LIMIT 1
```

**Reply by jeremyevans** [reply]
It doesn't make sense to use an association for this, as all instances of YoutubeAccount will have the same return value, and all instances of FacebookAccount will have the same return value.  You can do it if you really want to:

 ```ruby
many_to_one :network, dataset: ->(r){r.associated_class.where(slug: 'youtube')}
```

If the networks are not being modified, I would recommend constants:

```ruby
Network::Youtube = Network[slug: 'youtube']

class YoutubeAccount
  def network = Network::Youtube
end
```

**Reply by TSMMark** [reply]
Yeahh I understand it's not a usual kind of association. I'll just continue using static methods similar to your suggestion. Thanks!
